From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Wed, 12 Feb 2025 01:05:50 +0300
Subject: [PATCH] C2ME: Density Function Compiler

This functionality compiles density functions from world generation
datapacks (including vanilla generation) to JVM bytecode to increase
performance by allowing JVM JIT to better optimize the code.
All functions provided by vanilla are implemented.

About Density function: https://minecraft.wiki/w/Density_function

diff --git a/net/minecraft/util/CubicSpline.java b/net/minecraft/util/CubicSpline.java
index 110e50665256d1bf9fe49c7769be927ec50168a9..e87c5c061ead272da25bd9e77da3ba9bf15c55aa 100644
--- a/net/minecraft/util/CubicSpline.java
+++ b/net/minecraft/util/CubicSpline.java
@@ -270,7 +270,23 @@ public sealed interface CubicSpline<I> permits CubicSpline.Multipoint, CubicSpli
         }
 
         private static int findIntervalStart(final float[] locations, final float input) {
-            return Mth.binarySearch(0, locations.length, i -> input < locations[i]) - 1;
+            // DivineMC start - Density Function Compiler
+            int min = 0;
+            int i = locations.length;
+
+            while (i > 0) {
+                int j = i / 2;
+                int k = min + j;
+                if (input < locations[k]) {
+                    i = j;
+                } else {
+                    min = k + 1;
+                    i -= j + 1;
+                }
+            }
+
+            return min - 1;
+            // DivineMC end - Density Function Compiler
         }
 
         @VisibleForTesting
@@ -361,5 +377,27 @@ public sealed interface CubicSpline<I> permits CubicSpline.Multipoint, CubicSpli
                 );
             }
         }
+
+        // DivineMC start - Density Function Compiler
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Multipoint<?> that = (Multipoint<?>) o;
+            return java.util.Objects.equals(coordinate, that.coordinate()) && java.util.Arrays.equals(locations, that.locations()) && java.util.Objects.equals(values, that.values()) && java.util.Arrays.equals(derivatives, that.derivatives());
+        }
+
+        @Override
+        public int hashCode() {
+            int result = 1;
+
+            result = 31 * result + java.util.Objects.hashCode(coordinate);
+            result = 31 * result + java.util.Arrays.hashCode(locations);
+            result = 31 * result + java.util.Objects.hashCode(values);
+            result = 31 * result + java.util.Arrays.hashCode(derivatives);
+
+            return result;
+        }
+        // DivineMC end - Density Function Compiler
     }
 }
diff --git a/net/minecraft/world/level/levelgen/DensityFunction.java b/net/minecraft/world/level/levelgen/DensityFunction.java
index a8a7ef3ba147015d12d89e00c4f3e30da22546d7..62d3aa69ac8742805c2384669e8245b4ced01d35 100644
--- a/net/minecraft/world/level/levelgen/DensityFunction.java
+++ b/net/minecraft/world/level/levelgen/DensityFunction.java
@@ -111,6 +111,27 @@ public interface DensityFunction {
         public double maxValue() {
             return this.noise == null ? 2.0 : this.noise.maxValue();
         }
+
+        // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream MixinDensityFunctionNoise)
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            NoiseHolder that = (NoiseHolder) o;
+            return (this.noise != null || that.noise != null)
+                ? java.util.Objects.equals(this.noise, that.noise)
+                : java.util.Objects.equals(this.noiseData, that.noiseData);
+        }
+
+        @Override
+        public int hashCode() {
+            if (this.noise != null) {
+                return this.noise.hashCode();
+            } else {
+                return java.util.Objects.hashCode(this.noiseData);
+            }
+        }
+        // DivineMC end - Density Function Compiler
     }
 
     interface SimpleFunction extends DensityFunction {
diff --git a/net/minecraft/world/level/levelgen/DensityFunctions.java b/net/minecraft/world/level/levelgen/DensityFunctions.java
index 9f740bfbdddbdbdb729351574f1eb7169dc2e230..f1747cabca96f5ca47b985afab1fef68e1b3589b 100644
--- a/net/minecraft/world/level/levelgen/DensityFunctions.java
+++ b/net/minecraft/world/level/levelgen/DensityFunctions.java
@@ -278,7 +278,7 @@ public final class DensityFunctions {
         return new DensityFunctions.FindTopSurface(density, upperBound, lowerBound, stepSize);
     }
 
-    private record Ap2(
+    public record Ap2( // DivineMC - Density Function Compiler - private -> public (McToAst frontend registration)
         @Override DensityFunctions.TwoArgumentSimpleFunction.Type type,
         @Override DensityFunction argument1,
         @Override DensityFunction argument2,
@@ -302,12 +302,25 @@ public final class DensityFunctions {
             this.argument1.fillArray(output, contextProvider);
             switch (this.type) {
                 case ADD:
-                    double[] v2 = new double[output.length];
+                    // DivineMC start - Density Function Compiler - recycle the temporary array through DfcObjectCache (upstream MixinDFTBinaryOperation)
+                    double[] v2;
+                    com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache dfcObjectCache =
+                        contextProvider instanceof com.ishland.c2me.opts.dfc.common.ducks.IDfcObjectCacheCapable cacheCapable ? cacheCapable.c2me$getDfcObjectCache() : null;
+                    if (dfcObjectCache != null) {
+                        v2 = dfcObjectCache.getDoubleArray(output.length, false);
+                    } else {
+                        v2 = new double[output.length];
+                    }
                     this.argument2.fillArray(v2, contextProvider);
 
                     for (int i = 0; i < output.length; i++) {
                         output[i] += v2[i];
                     }
+
+                    if (dfcObjectCache != null) {
+                        dfcObjectCache.recycle(v2);
+                    }
+                    // DivineMC end - Density Function Compiler
                     break;
                 case MUL:
                     for (int i = 0; i < output.length; i++) {
@@ -339,7 +352,7 @@ public final class DensityFunctions {
         }
     }
 
-    enum BeardifierMarker implements DensityFunctions.BeardifierOrMarker {
+    public enum BeardifierMarker implements DensityFunctions.BeardifierOrMarker { // DivineMC - Density Function Compiler - protected -> public
         INSTANCE;
 
         @Override
@@ -493,12 +506,27 @@ public final class DensityFunctions {
         }
     }
 
-    protected static final class EndIslandDensityFunction implements DensityFunction.SimpleFunction {
+    public static final class EndIslandDensityFunction implements DensityFunction.SimpleFunction { // DivineMC - Density Function Compiler - protected -> public
         public static final KeyDispatchDataCodec<DensityFunctions.EndIslandDensityFunction> CODEC = KeyDispatchDataCodec.of(
             MapCodec.unit(new DensityFunctions.EndIslandDensityFunction(0L))
         );
         private static final float ISLAND_THRESHOLD = -0.9F;
         private final SimplexNoise islandNoise;
+
+        // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream MixinDFTEndIslands)
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            EndIslandDensityFunction that = (EndIslandDensityFunction) o;
+            return java.util.Objects.equals(this.islandNoise, that.islandNoise);
+        }
+
+        @Override
+        public int hashCode() {
+            return java.util.Objects.hashCode(this.islandNoise);
+        }
+        // DivineMC end - Density Function Compiler
         // Paper start - Perf: Optimize end generation
         private static final class NoiseCache {
             public long[] keys = new long[8192];
@@ -828,7 +856,75 @@ public final class DensityFunctions {
         }
     }
 
-    record Marker(@Override DensityFunctions.Marker.Type type, @Override DensityFunction wrapped) implements DensityFunctions.MarkerOrMarked {
+    // DivineMC start - Density Function Compiler - Marker record converted to a class implementing IFastCacheLike (upstream MixinDFTWrapping)
+    public static final class Marker implements com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike, MarkerOrMarked {
+        private final Type type;
+        private final DensityFunction wrapped;
+
+        public Marker(Type type, DensityFunction wrapped) {
+            this.type = type;
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (o == this) return true;
+            if (o == null || o.getClass() != this.getClass()) return false;
+            Marker that = (Marker) o;
+            return this.type == that.type && java.util.Objects.equals(this.wrapped, that.wrapped);
+        }
+
+        @Override
+        public int hashCode() {
+            return java.util.Objects.hash(this.type, this.wrapped);
+        }
+
+        @Override
+        public double c2me$getCached(int x, int y, int z, com.ishland.c2me.opts.dfc.common.ast.EvalType evalType) {
+            return Double.longBitsToDouble(CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(double[] res, int[] x, int[] y, int[] z, com.ishland.c2me.opts.dfc.common.ast.EvalType evalType) {
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(int x, int y, int z, com.ishland.c2me.opts.dfc.common.ast.EvalType evalType, double cached) {
+            // nop
+        }
+
+        @Override
+        public void c2me$cache(double[] res, int[] x, int[] y, int[] z, com.ishland.c2me.opts.dfc.common.ast.EvalType evalType) {
+            // nop
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return this.type != Type.Interpolated;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "wrapping, " + this.type.getSerializedName();
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.wrapped;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(DensityFunction delegate) {
+            return new DensityFunctions.Marker(this.type(), delegate);
+        }
+
+        @Override
+        public DensityFunction mapAll(Visitor visitor) {
+            return visitor.apply(this.c2me$withDelegate(this.wrapped().mapAll(visitor)));
+        }
+        // DivineMC end - Density Function Compiler
+
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
             return this.wrapped.compute(context);
@@ -849,7 +945,19 @@ public final class DensityFunctions {
             return this.type == DensityFunctions.Marker.Type.BlendDensity ? Double.POSITIVE_INFINITY : this.wrapped.maxValue();
         }
 
-        public enum Type implements StringRepresentable {
+        // DivineMC start - Density Function Compiler
+        @Override
+        public Type type() {
+            return type;
+        }
+
+        @Override
+        public DensityFunction wrapped() {
+            return wrapped;
+        }
+        // DivineMC end - Density Function Compiler
+
+        public enum Type implements StringRepresentable { // DivineMC - Density Function Compiler - make this public
             Interpolated("interpolated"),
             FlatCache("flat_cache"),
             Cache2D("cache_2d"),
@@ -889,7 +997,7 @@ public final class DensityFunctions {
         }
     }
 
-    private record MulOrAdd(
+    public record MulOrAdd( // DivineMC - Density Function Compiler - private -> public (McToAst frontend registration)
         DensityFunctions.MulOrAdd.Type specificType, @Override DensityFunction input, @Override double minValue, @Override double maxValue, double argument
     ) implements DensityFunctions.TwoArgumentSimpleFunction, DensityFunctions.PureTransformer {
         @Override
diff --git a/net/minecraft/world/level/levelgen/NoiseChunk.java b/net/minecraft/world/level/levelgen/NoiseChunk.java
index 16a656a50ba2c39ca4f9facf7d71bf697ebc7658..11c97f273879be95f47884ab560ae88f7ea04c9f 100644
--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -4,6 +4,7 @@ import com.google.common.collect.Lists;
 import it.unimi.dsi.fastutil.longs.Long2IntMap;
 import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
 import java.util.ArrayList;
+import java.util.Arrays; // DivineMC - Density Function Compiler
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -20,8 +21,20 @@ import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.levelgen.blending.Blender;
 import net.minecraft.world.level.levelgen.material.MaterialRuleList;
 import org.jspecify.annotations.Nullable;
-
-public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunction.ContextProvider {
+// DivineMC start - Density Function Compiler
+import com.ishland.c2me.opts.dfc.common.ast.EvalType;
+import com.ishland.c2me.opts.dfc.common.ducks.ICoordinatesFilling;
+import com.ishland.c2me.opts.dfc.common.ducks.IDfcObjectCacheCapable;
+import com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike;
+import com.ishland.c2me.opts.dfc.common.ducks.IPreloadedCoordinates;
+import com.ishland.c2me.opts.dfc.common.ducks.NoiseRouterExtension;
+import com.ishland.c2me.opts.dfc.common.gen.DelegatingBlendingAwareVisitor;
+import com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache;
+import com.ishland.c2me.opts.dfc.common.gen.jvm.vif.EachApplierVanillaInterface;
+import com.ishland.c2me.opts.dfc.common.gen.jvm.vif.NoisePosVanillaInterface;
+// DivineMC end - Density Function Compiler
+
+public class NoiseChunk implements DensityFunction.ContextProvider, DensityFunction.FunctionContext, IDfcObjectCacheCapable, ICoordinatesFilling, IPreloadedCoordinates { // DivineMC - Density Function Compiler
     private final int cellCountXZ;
     private final int cellCountY;
     private final int cellNoiseMinY;
@@ -57,7 +70,88 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
     private long interpolationCounter;
     private long arrayInterpolationCounter;
     private int arrayIndex;
-    private final DensityFunction.ContextProvider sliceFillingContextProvider = new DensityFunction.ContextProvider() {
+    // DivineMC start - Density Function Compiler
+    private final DfcObjectCache c2me$dfcObjectCache = new DfcObjectCache.Impl();
+
+    private int c2me$oldStartBlockX = Integer.MAX_VALUE;
+    private int c2me$oldStartBlockY = Integer.MAX_VALUE;
+    private int c2me$oldStartBlockZ = Integer.MAX_VALUE;
+    private int[] c2me$cachedXArray;
+    private int[] c2me$cachedYArray;
+    private int[] c2me$cachedZArray;
+
+    @Override
+    public DfcObjectCache c2me$getDfcObjectCache() {
+        return java.util.Objects.requireNonNull(this.c2me$dfcObjectCache, "broken initialization");
+    }
+
+    @Override
+    public void c2me$fillCoordinates(int[] x, int[] y, int[] z) {
+        int index = 0;
+        for (int i = this.cellHeight - 1; i >= 0; i--) {
+            int blockY = this.cellStartBlockY + i;
+            for (int j = 0; j < this.cellWidth; j++) {
+                int blockX = this.cellStartBlockX + j;
+                for (int k = 0; k < this.cellWidth; k++) {
+                    int blockZ = this.cellStartBlockZ + k;
+
+                    x[index] = blockX;
+                    y[index] = blockY;
+                    z[index] = blockZ;
+
+                    index++;
+                }
+            }
+        }
+    }
+
+    private void c2me$reloadCachedArrays() {
+        if (this.c2me$cachedXArray == null || this.c2me$cachedYArray == null || this.c2me$cachedZArray == null) {
+            int length = this.cellWidth * this.cellWidth * this.cellHeight;
+            DfcObjectCache dfcObjectCache = this.c2me$getDfcObjectCache();
+            this.c2me$cachedXArray = dfcObjectCache.getIntArray(length, false);
+            this.c2me$cachedYArray = dfcObjectCache.getIntArray(length, false);
+            this.c2me$cachedZArray = dfcObjectCache.getIntArray(length, false);
+        }
+        if (this.c2me$oldStartBlockX != this.cellStartBlockX || this.c2me$oldStartBlockY != this.cellStartBlockY || this.c2me$oldStartBlockZ != this.cellStartBlockZ) {
+            this.c2me$fillCoordinates(this.c2me$cachedXArray, this.c2me$cachedYArray, this.c2me$cachedZArray);
+            this.c2me$oldStartBlockX = this.cellStartBlockX;
+            this.c2me$oldStartBlockY = this.cellStartBlockY;
+            this.c2me$oldStartBlockZ = this.cellStartBlockZ;
+        }
+    }
+
+    @Override
+    public int[] c2me$getXArray() {
+        this.c2me$reloadCachedArrays();
+        return this.c2me$cachedXArray;
+    }
+
+    @Override
+    public int[] c2me$getYArray() {
+        this.c2me$reloadCachedArrays();
+        return this.c2me$cachedYArray;
+    }
+
+    @Override
+    public int[] c2me$getZArray() {
+        this.c2me$reloadCachedArrays();
+        return this.c2me$cachedZArray;
+    }
+
+    private @org.jetbrains.annotations.NotNull DelegatingBlendingAwareVisitor c2me$getDelegatingBlendingAwareVisitor(DensityFunction.Visitor visitor) {
+        return new DelegatingBlendingAwareVisitor(visitor, !this.blender.isEmpty()); // DivineMC - 26.2: FunctionContext.getBlender() removed
+    }
+
+    private DensityFunction.Visitor modifyVisitor1(DensityFunction.Visitor visitor) {
+        return c2me$getDelegatingBlendingAwareVisitor(visitor);
+    }
+
+    private DensityFunction.Visitor modifyVisitor2(DensityFunction.Visitor visitor) {
+        return c2me$getDelegatingBlendingAwareVisitor(visitor);
+    }
+
+    public class NoiseChunkSliceFillingContextProvider implements DensityFunction.ContextProvider, IDfcObjectCacheCapable, ICoordinatesFilling {
         @Override
         public DensityFunction.FunctionContext forIndex(final int cellYIndex) {
             NoiseChunk.this.cellStartBlockY = (cellYIndex + NoiseChunk.this.cellNoiseMinY) * NoiseChunk.this.cellHeight;
@@ -77,7 +171,23 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
                 output[cellYIndex] = function.compute(NoiseChunk.this);
             }
         }
-    };
+
+        @Override
+        public DfcObjectCache c2me$getDfcObjectCache() {
+            return (NoiseChunk.this).c2me$getDfcObjectCache();
+        }
+
+        @Override
+        public void c2me$fillCoordinates(int[] x, int[] y, int[] z) {
+            for (int i = 0; i < (NoiseChunk.this).cellCountY + 1; i++) {
+                x[i] = (NoiseChunk.this).cellStartBlockX + (NoiseChunk.this).inCellX;
+                y[i] = (i + (NoiseChunk.this).cellNoiseMinY) * (NoiseChunk.this).cellHeight;
+                z[i] = (NoiseChunk.this).cellStartBlockZ + (NoiseChunk.this).inCellZ;
+            }
+        }
+    }
+    private final DensityFunction.ContextProvider sliceFillingContextProvider = new NoiseChunkSliceFillingContextProvider();
+    // DivineMC end - Density Function Compiler
 
     public static NoiseChunk forChunk(
         final ChunkAccess chunk,
@@ -140,7 +250,7 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         }
 
         NoiseRouter router = randomState.router();
-        NoiseRouter wrappedRouter = router.mapAll(this::wrap);
+        NoiseRouter wrappedRouter = router.mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor1(this::wrap) : this::wrap); // DivineMC - Density Function Compiler
         this.preliminarySurfaceLevel = wrappedRouter.preliminarySurfaceLevel();
         if (!settings.isAquifersEnabled()) {
             this.aquifer = Aquifer.createDisabled(globalFluidPicker);
@@ -153,10 +263,21 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         }
 
         List<NoiseChunk.BlockStateFiller> builder = new ArrayList<>();
-        DensityFunction fullNoiseValue = DensityFunctions.cacheAllInCell(
-                DensityFunctions.add(wrappedRouter.finalDensity(), DensityFunctions.BeardifierMarker.INSTANCE)
-            )
-            .mapAll(this::wrap);
+        // DivineMC start - Density Function Compiler - use the precompiled finalDensity+Beardifier when available (upstream wrapFinalFinalDensity0-3);
+        // it was already propagated through the wrap visitor by NoiseRouter#mapAll above, so it is used as-is
+        DensityFunction fullNoiseValue;
+        DensityFunction precompiledFinalFinalDensity = org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler
+            ? ((NoiseRouterExtension) wrappedRouter).c2me$getFinalFinalDensity()
+            : null;
+        if (precompiledFinalFinalDensity != null) {
+            fullNoiseValue = precompiledFinalFinalDensity;
+        } else {
+            fullNoiseValue = DensityFunctions.cacheAllInCell(
+                    DensityFunctions.add(wrappedRouter.finalDensity(), DensityFunctions.BeardifierMarker.INSTANCE)
+                )
+                .mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap);
+        }
+        // DivineMC end - Density Function Compiler
         this.fullNoiseDensity = fullNoiseValue;
         builder.add(context -> this.aquifer.computeSubstance(context, fullNoiseValue.compute(context)));
         if (settings.oreVeinsEnabled()) {
@@ -168,12 +289,14 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
 
     protected Climate.Sampler cachedClimateSampler(final NoiseRouter noises, final List<Climate.ParameterPoint> spawnTarget) {
         return new Climate.Sampler(
-            noises.temperature().mapAll(this::wrap),
-            noises.vegetation().mapAll(this::wrap),
-            noises.continents().mapAll(this::wrap),
-            noises.erosion().mapAll(this::wrap),
-            noises.depth().mapAll(this::wrap),
-            noises.ridges().mapAll(this::wrap),
+            // DivineMC start - Density Function Compiler
+            noises.temperature().mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap),
+            noises.vegetation().mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap),
+            noises.continents().mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap),
+            noises.erosion().mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap),
+            noises.depth().mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap),
+            noises.ridges().mapAll(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler ? modifyVisitor2(this::wrap) : this::wrap),
+            // DivineMC end - Density Function Compiler
             spawnTarget
         );
     }
@@ -311,10 +434,21 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         this.fillingCell = false;
     }
 
+    // DivineMC start - Density Function Compiler - iterate interpolators over a cached array (upstream MixinChunkNoiseSampler overwrites)
+    private NoiseChunk.NoiseInterpolator @Nullable [] c2me$interpolatorsArray;
+
+    private NoiseChunk.NoiseInterpolator[] c2me$initInterpolatorsArray() {
+        NoiseChunk.NoiseInterpolator[] interpolatorsArray = this.c2me$interpolatorsArray;
+        if (interpolatorsArray == null) {
+            interpolatorsArray = this.c2me$interpolatorsArray = this.interpolators.toArray(NoiseChunk.NoiseInterpolator[]::new);
+        }
+        return interpolatorsArray;
+    }
+
     public void updateForY(final int posY, final double factorY) {
         this.inCellY = posY - this.cellStartBlockY;
 
-        for (NoiseChunk.NoiseInterpolator i : this.interpolators) {
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) {
             i.updateForY(factorY);
         }
     }
@@ -322,7 +456,7 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
     public void updateForX(final int posX, final double factorX) {
         this.inCellX = posX - this.cellStartBlockX;
 
-        for (NoiseChunk.NoiseInterpolator i : this.interpolators) {
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) {
             i.updateForX(factorX);
         }
     }
@@ -331,7 +465,7 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         this.inCellZ = posZ - this.cellStartBlockZ;
         this.interpolationCounter++;
 
-        for (NoiseChunk.NoiseInterpolator i : this.interpolators) {
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) {
             i.updateForZ(factorZ);
         }
     }
@@ -345,8 +479,11 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
     }
 
     public void swapSlices() {
-        this.interpolators.forEach(NoiseChunk.NoiseInterpolator::swapSlices);
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) {
+            i.swapSlices();
+        }
     }
+    // DivineMC end - Density Function Compiler
 
     public Aquifer aquifer() {
         return this.aquifer;
@@ -377,8 +514,19 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
     }
 
     private DensityFunction wrapNew(final DensityFunction function) {
+        // DivineMC start - Density Function Compiler
+        if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler) {
+            if (this.interpolating && function instanceof DensityFunctions.Marker) {
+                throw new IllegalStateException("Cannot create more wrapping during interpolation loop");
+            }
+        }
+        // DivineMC end - Density Function Compiler
         return switch (function) {
-            case DensityFunctions.Marker(DensityFunctions.Marker.Type type, DensityFunction wrapped) -> {
+            case DensityFunctions.Marker marker -> { // DivineMC - Density Function Compiler - Marker is a class, not a record
+                // DivineMC start - Density Function Compiler
+                final DensityFunctions.Marker.Type type = marker.type();
+                final DensityFunction wrapped = marker.wrapped();
+                // DivineMC end - Density Function Compiler
                 switch (type) {
                     case Interpolated:
                         yield new NoiseChunk.NoiseInterpolator(wrapped);
@@ -531,10 +679,58 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         @Nullable BlockState calculate(final DensityFunction.FunctionContext context);
     }
 
-    private static class Cache2D implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
+    public static class Cache2D implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, IFastCacheLike { // DivineMC - Density Function Compiler - private -> public, IFastCacheLike (upstream MixinChunkNoiseSamplerCache2D)
         private DensityFunction function;
         private long lastPos2D = ChunkPos.INVALID_CHUNK_POS;
         private double lastValue;
+        // DivineMC start - Density Function Compiler
+        @Override
+        public double c2me$getCached(int x, int y, int z, EvalType evalType) {
+            long l = ChunkPos.pack(x, z);
+            if (this.lastPos2D == l) {
+                return this.lastValue;
+            } else {
+                return Double.longBitsToDouble(CACHE_MISS_NAN_BITS);
+            }
+        }
+
+        @Override
+        public boolean c2me$getCached(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(int x, int y, int z, EvalType evalType, double cached) {
+            this.lastPos2D = ChunkPos.pack(x, z);
+            this.lastValue = cached;
+        }
+
+        @Override
+        public void c2me$cache(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            // nop
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "cache_2d";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.function;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(DensityFunction delegate) {
+            this.function = delegate;
+            return this;
+        }
+        // DivineMC end - Density Function Compiler
 
         private Cache2D(final DensityFunction function) {
             this.function = function;
@@ -571,9 +767,93 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         }
     }
 
-    private class CacheAllInCell implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
-        private final DensityFunction noiseFiller;
+    public class CacheAllInCell implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, IFastCacheLike { // DivineMC - Density Function Compiler - private -> public, IFastCacheLike (upstream MixinChunkNoiseSamplerCellCache)
+        private DensityFunction noiseFiller; // DivineMC - Density Function Compiler - remove final
         private final double[] values;
+        // DivineMC start - Density Function Compiler
+        @Override
+        public double c2me$getCached(int x, int y, int z, EvalType evalType) {
+            if (evalType == EvalType.INTERPOLATION && NoiseChunk.this.interpolating) {
+                int cellBlockX = x - NoiseChunk.this.cellStartBlockX;
+                int cellBlockY = y - NoiseChunk.this.cellStartBlockY;
+                int cellBlockZ = z - NoiseChunk.this.cellStartBlockZ;
+                int horizontalCellBlockCount = NoiseChunk.this.cellWidth;
+                int verticalCellBlockCount = NoiseChunk.this.cellHeight;
+                if (cellBlockX >= 0 &&
+                    cellBlockY >= 0 &&
+                    cellBlockZ >= 0 &&
+                    cellBlockX < horizontalCellBlockCount &&
+                    cellBlockY < verticalCellBlockCount &&
+                    cellBlockZ < horizontalCellBlockCount) {
+                    return this.values[((verticalCellBlockCount - 1 - cellBlockY) * horizontalCellBlockCount + cellBlockX)
+                        * horizontalCellBlockCount
+                        + cellBlockZ];
+                }
+            }
+
+            return Double.longBitsToDouble(CACHE_MISS_NAN_BITS); // DivineMC - fix upstream bug: long constant was implicitly widened to a bogus double
+        }
+
+        @Override
+        public boolean c2me$getCached(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            if (evalType == EvalType.INTERPOLATION && NoiseChunk.this.interpolating) {
+                int startBlockX = NoiseChunk.this.cellStartBlockX;
+                int startBlockY = NoiseChunk.this.cellStartBlockY;
+                int startBlockZ = NoiseChunk.this.cellStartBlockZ;
+                int horizontalCellBlockCount = NoiseChunk.this.cellWidth;
+                int verticalCellBlockCount = NoiseChunk.this.cellHeight;
+                for (int i = 0; i < res.length; i++) {
+                    int cellBlockX = x[i] - startBlockX;
+                    int cellBlockY = y[i] - startBlockY;
+                    int cellBlockZ = z[i] - startBlockZ;
+                    if (cellBlockX >= 0 &&
+                        cellBlockY >= 0 &&
+                        cellBlockZ >= 0 &&
+                        cellBlockX < horizontalCellBlockCount &&
+                        cellBlockY < verticalCellBlockCount &&
+                        cellBlockZ < horizontalCellBlockCount) {
+                        res[i] = this.values[((verticalCellBlockCount - 1 - cellBlockY) * horizontalCellBlockCount + cellBlockX) * horizontalCellBlockCount + cellBlockZ];
+                    } else {
+                        return false; // partial hit possible
+                    }
+                }
+                return true; // DivineMC - fix upstream bug: a full scan without misses is a complete hit
+            }
+
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(int x, int y, int z, EvalType evalType, double cached) {
+            // nop
+        }
+
+        @Override
+        public void c2me$cache(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            // nop
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "cache_all_in_cell";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.noiseFiller;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(DensityFunction delegate) {
+            this.noiseFiller = delegate;
+            return this;
+        }
+        // DivineMC end - Density Function Compiler
 
         private CacheAllInCell(final DensityFunction noiseFiller) {
             this.noiseFiller = noiseFiller;
@@ -584,6 +864,24 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
             if (context != NoiseChunk.this) {
+                // DivineMC start - Density Function Compiler - serve compiled-code lookups from the cell cache (upstream MixinChunkNoiseSamplerCellCache#wrapSample)
+                if (context instanceof NoisePosVanillaInterface vif && vif.getType() == EvalType.INTERPOLATION && NoiseChunk.this.interpolating) {
+                    int cellBlockX = context.blockX() - NoiseChunk.this.cellStartBlockX;
+                    int cellBlockY = context.blockY() - NoiseChunk.this.cellStartBlockY;
+                    int cellBlockZ = context.blockZ() - NoiseChunk.this.cellStartBlockZ;
+                    int horizontalCellBlockCount = NoiseChunk.this.cellWidth;
+                    int verticalCellBlockCount = NoiseChunk.this.cellHeight;
+                    return cellBlockX >= 0
+                        && cellBlockY >= 0
+                        && cellBlockZ >= 0
+                        && cellBlockX < horizontalCellBlockCount
+                        && cellBlockY < verticalCellBlockCount
+                        && cellBlockZ < horizontalCellBlockCount
+                        ? this.values[((verticalCellBlockCount - 1 - cellBlockY) * horizontalCellBlockCount + cellBlockX)
+                                      * horizontalCellBlockCount + cellBlockZ]
+                        : this.noiseFiller.compute(context);
+                }
+                // DivineMC end - Density Function Compiler
                 return this.noiseFiller.compute(context);
             }
 
@@ -615,12 +913,93 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         }
     }
 
-    private class CacheOnce implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
+    public class CacheOnce implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, IFastCacheLike { // DivineMC - Density Function Compiler - private -> public, IFastCacheLike (upstream MixinChunkNoiseSamplerCacheOnce)
         private DensityFunction function;
         private long lastCounter;
         private long lastArrayCounter;
         private double lastValue;
         private double @Nullable [] lastArray;
+        // DivineMC start - Density Function Compiler
+        private double c2me$lastValue = Double.NaN;
+        private int c2me$lastX = Integer.MIN_VALUE;
+        private int c2me$lastY = Integer.MIN_VALUE;
+        private int c2me$lastZ = Integer.MIN_VALUE;
+
+        private int[] c2me$lastXa;
+        private int[] c2me$lastYa;
+        private int[] c2me$lastZa;
+        private double[] c2me$lastValuea;
+
+        @Override
+        public double c2me$getCached(int x, int y, int z, EvalType evalType) {
+            if (c2me$lastValuea != null) {
+                for (int i = 0; i < this.c2me$lastValuea.length; i ++) {
+                    if (c2me$lastXa[i] == x && c2me$lastYa[i] == y && c2me$lastZa[i] == z) {
+                        return c2me$lastValuea[i];
+                    }
+                }
+            }
+            if (!Double.isNaN(c2me$lastValue) && c2me$lastX == x && c2me$lastY == y && c2me$lastZ == z) {
+                return c2me$lastValue;
+            }
+
+            return Double.longBitsToDouble(CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            if (c2me$lastValuea != null && Arrays.equals(y, c2me$lastYa) && Arrays.equals(x, c2me$lastXa) && Arrays.equals(z, c2me$lastZa)) {
+                System.arraycopy(c2me$lastValuea, 0, res, 0, c2me$lastValuea.length);
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public void c2me$cache(int x, int y, int z, EvalType evalType, double cached) {
+            c2me$lastValue = cached;
+            c2me$lastX = x;
+            c2me$lastY = y;
+            c2me$lastZ = z;
+        }
+
+        @Override
+        public void c2me$cache(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            if (c2me$lastValuea != null && this.c2me$lastValuea.length == res.length) {
+                System.arraycopy(res, 0, this.c2me$lastValuea, 0, this.c2me$lastValuea.length);
+                System.arraycopy(x, 0, this.c2me$lastXa, 0, this.c2me$lastValuea.length);
+                System.arraycopy(y, 0, this.c2me$lastYa, 0, this.c2me$lastValuea.length);
+                System.arraycopy(z, 0, this.c2me$lastZa, 0, this.c2me$lastValuea.length);
+            } else {
+                this.c2me$lastValuea = Arrays.copyOf(res, res.length);
+                this.c2me$lastXa = Arrays.copyOf(x, x.length);
+                this.c2me$lastYa = Arrays.copyOf(y, y.length);
+                this.c2me$lastZa = Arrays.copyOf(z, z.length);
+            }
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "cache_once";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.function;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(DensityFunction delegate) {
+            this.function = delegate;
+            return this;
+        }
+        // DivineMC end - Density Function Compiler
 
         private CacheOnce(final DensityFunction function) {
             this.function = function;
@@ -628,26 +1007,76 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
 
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
-            if (context != NoiseChunk.this) {
-                return this.function.compute(context);
+            // DivineMC start - Density Function Compiler - vanilla path for NoiseChunk contexts, coordinate cache otherwise (upstream MixinChunkNoiseSamplerCacheOnce#wrapSample)
+            if (context instanceof NoiseChunk) {
+                if (context != NoiseChunk.this) {
+                    return this.function.compute(context);
+                }
+
+                if (this.lastArray != null
+                    && this.lastArrayCounter == NoiseChunk.this.arrayInterpolationCounter) {
+                    return this.lastArray[NoiseChunk.this.arrayIndex];
+                }
+
+                if (this.lastCounter == NoiseChunk.this.interpolationCounter) {
+                    return this.lastValue;
+                }
+
+                this.lastCounter = NoiseChunk.this.interpolationCounter;
+                double value = this.function.compute(context);
+                this.lastValue = value;
+                return value;
             }
 
-            if (this.lastArray != null && this.lastArrayCounter == NoiseChunk.this.arrayInterpolationCounter) {
-                return this.lastArray[NoiseChunk.this.arrayIndex];
+            int blockX = context.blockX();
+            int blockY = context.blockY();
+            int blockZ = context.blockZ();
+
+            if (c2me$lastValuea != null) {
+                for (int i = 0; i < this.c2me$lastValuea.length; i++) {
+                    if (c2me$lastXa[i] == blockX
+                        && c2me$lastYa[i] == blockY
+                        && c2me$lastZa[i] == blockZ) {
+                        return c2me$lastValuea[i];
+                    }
+                }
             }
 
-            if (this.lastCounter == NoiseChunk.this.interpolationCounter) {
-                return this.lastValue;
+            if (!Double.isNaN(c2me$lastValue)
+                && c2me$lastX == blockX
+                && c2me$lastY == blockY
+                && c2me$lastZ == blockZ) {
+                return c2me$lastValue;
             }
 
-            this.lastCounter = NoiseChunk.this.interpolationCounter;
-            double value = this.function.compute(context);
-            this.lastValue = value;
-            return value;
+            double sample = this.function.compute(context);
+
+            c2me$lastValue = sample;
+            c2me$lastX = blockX;
+            c2me$lastY = blockY;
+            c2me$lastZ = blockZ;
+
+            return sample;
+            // DivineMC end - Density Function Compiler
         }
 
         @Override
         public void fillArray(final double[] output, final DensityFunction.ContextProvider contextProvider) {
+            // DivineMC start - Density Function Compiler - divert compiled-code (VIF) calls into the coordinate array cache (upstream MixinChunkNoiseSamplerCacheOnce#wrapFill);
+            // every vanilla provider (NoiseChunk, slice filler, ...) keeps the vanilla lastArray caching below
+            if (contextProvider instanceof EachApplierVanillaInterface ap) {
+                if (c2me$lastValuea != null && Arrays.equals(ap.getY(), c2me$lastYa) && Arrays.equals(ap.getX(), c2me$lastXa) && Arrays.equals(ap.getZ(), c2me$lastZa)) {
+                    System.arraycopy(c2me$lastValuea, 0, output, 0, c2me$lastValuea.length);
+                } else {
+                    this.function.fillArray(output, contextProvider);
+                    this.c2me$lastValuea = Arrays.copyOf(output, output.length);
+                    this.c2me$lastXa = ap.getX();
+                    this.c2me$lastYa = ap.getY();
+                    this.c2me$lastZa = ap.getZ();
+                }
+                return;
+            }
+            // DivineMC end - Density Function Compiler
             if (this.lastArray != null && this.lastArrayCounter == NoiseChunk.this.arrayInterpolationCounter) {
                 System.arraycopy(this.lastArray, 0, output, 0, output.length);
             } else {
@@ -655,7 +1084,7 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
                 if (this.lastArray != null && this.lastArray.length == output.length) {
                     System.arraycopy(output, 0, this.lastArray, 0, output.length);
                 } else {
-                    this.lastArray = (double[])output.clone();
+                    this.lastArray = (double[]) output.clone();
                 }
 
                 this.lastArrayCounter = NoiseChunk.this.arrayInterpolationCounter;
@@ -673,10 +1102,73 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         }
     }
 
-    private class FlatCache implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
+    public class FlatCache implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, IFastCacheLike { // DivineMC - Density Function Compiler - private -> public, IFastCacheLike (upstream MixinChunkNoiseSamplerFlatCache)
         private DensityFunction noiseFiller;
         private final double[] values;
         private final int sizeXZ;
+        // DivineMC start - Density Function Compiler
+        @Override
+        public double c2me$getCached(int x, int y, int z, EvalType evalType) {
+            int i = QuartPos.fromBlock(x);
+            int j = QuartPos.fromBlock(z);
+            int k = i - (NoiseChunk.this).firstNoiseX;
+            int l = j - (NoiseChunk.this).firstNoiseZ;
+            int m = this.sizeXZ; // DivineMC - fix old port bug: bound/stride must be sizeXZ, not values.length
+            if (k >= 0 && l >= 0 && k < m && l < m) {
+                return this.values[k + l * m];
+            } else {
+                return Double.longBitsToDouble(CACHE_MISS_NAN_BITS);
+            }
+        }
+
+        @Override
+        public boolean c2me$getCached(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            for (int i = 0; i < res.length; i ++) {
+                int i1 = QuartPos.fromBlock(x[i]);
+                int j1 = QuartPos.fromBlock(z[i]);
+                int k = i1 - (NoiseChunk.this).firstNoiseX;
+                int l = j1 - (NoiseChunk.this).firstNoiseZ;
+                int m = this.sizeXZ; // DivineMC - fix old port bug: bound/stride must be sizeXZ, not values.length
+                if (k >= 0 && l >= 0 && k < m && l < m) {
+                    res[i] = this.values[k + l * m];
+                } else {
+                    return false; // partial hit possible
+                }
+            }
+            return true;
+        }
+
+        @Override
+        public void c2me$cache(int x, int y, int z, EvalType evalType, double cached) {
+            // nop
+        }
+
+        @Override
+        public void c2me$cache(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            // nop
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "flat_cache";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.noiseFiller;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(DensityFunction delegate) {
+            this.noiseFiller = delegate;
+            return this;
+        }
+        // DivineMC end - Density Function Compiler
 
         private FlatCache(final DensityFunction noiseFiller, final boolean fill) {
             this.noiseFiller = noiseFiller;
@@ -735,7 +1227,7 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         }
     }
 
-    public class NoiseInterpolator implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
+    public class NoiseInterpolator implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, IFastCacheLike { // DivineMC - Density Function Compiler
         private double[][] slice0;
         private double[][] slice1;
         private DensityFunction noiseFiller;
@@ -754,6 +1246,114 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         private double valueZ0;
         private double valueZ1;
         private double value;
+        // DivineMC start - Density Function Compiler
+        @Override
+        public double c2me$getCached(int x, int y, int z, EvalType evalType) {
+            if (evalType == EvalType.INTERPOLATION) {
+                boolean isInInterpolationLoop = (NoiseChunk.this).interpolating;
+                if (isInInterpolationLoop) {
+                    if ((NoiseChunk.this).fillingCell) {
+                        int startBlockX = (NoiseChunk.this).cellStartBlockX;
+                        int startBlockY = (NoiseChunk.this).cellStartBlockY;
+                        int startBlockZ = (NoiseChunk.this).cellStartBlockZ;
+                        int horizontalCellBlockCount = (NoiseChunk.this).cellWidth;
+                        int verticalCellBlockCount = (NoiseChunk.this).cellHeight;
+                        int cellBlockX = x - startBlockX;
+                        int cellBlockY = y - startBlockY;
+                        int cellBlockZ = z - startBlockZ;
+                        return Mth.lerp3(
+                            (double) cellBlockX / (double) horizontalCellBlockCount,
+                            (double) cellBlockY / (double) verticalCellBlockCount,
+                            (double) cellBlockZ / (double) horizontalCellBlockCount,
+                            this.noise000,
+                            this.noise100,
+                            this.noise010,
+                            this.noise110,
+                            this.noise001,
+                            this.noise101,
+                            this.noise011,
+                            this.noise111
+                        );
+                    } else {
+                        return this.value;
+                    }
+                }
+            }
+
+            return Double.longBitsToDouble(CACHE_MISS_NAN_BITS); // DivineMC - fix upstream bug: long constant was implicitly widened to a bogus double
+        }
+
+        @Override
+        public boolean c2me$getCached(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            if (evalType == EvalType.INTERPOLATION) {
+                boolean isInInterpolationLoop = (NoiseChunk.this).interpolating;
+                if (isInInterpolationLoop) {
+                    if ((NoiseChunk.this).fillingCell) {
+                        int startBlockX = (NoiseChunk.this).cellStartBlockX;
+                        int startBlockY = (NoiseChunk.this).cellStartBlockY;
+                        int startBlockZ = (NoiseChunk.this).cellStartBlockZ;
+                        double horizontalCellBlockCount = (NoiseChunk.this).cellWidth;
+                        double verticalCellBlockCount = (NoiseChunk.this).cellHeight;
+                        for (int i = 0; i < res.length; i ++) {
+                            int cellBlockX = x[i] - startBlockX;
+                            int cellBlockY = y[i] - startBlockY;
+                            int cellBlockZ = z[i] - startBlockZ;
+                            res[i] = Mth.lerp3(
+                                (double)cellBlockX / horizontalCellBlockCount,
+                                (double)cellBlockY / verticalCellBlockCount,
+                                (double)cellBlockZ / horizontalCellBlockCount,
+                                this.noise000,
+                                this.noise100,
+                                this.noise010,
+                                this.noise110,
+                                this.noise001,
+                                this.noise101,
+                                this.noise011,
+                                this.noise111
+                            );
+                        }
+                        return true;
+                    } else {
+                        Arrays.fill(res, this.value);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(int x, int y, int z, EvalType evalType, double cached) {
+            // nop
+        }
+
+        @Override
+        public void c2me$cache(double[] res, int[] x, int[] y, int[] z, EvalType evalType) {
+            // nop
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "interpolated";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.noiseFiller;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(DensityFunction delegate) {
+            this.noiseFiller = delegate;
+            return this;
+        }
+        // DivineMC end - Density Function Compiler
 
         private NoiseInterpolator(final DensityFunction noiseFiller) {
             this.noiseFiller = noiseFiller;
@@ -804,15 +1404,37 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
             if (context != NoiseChunk.this) {
+                // DivineMC start - Density Function Compiler - serve compiled-code (VIF) lookups from the interpolator state (upstream MixinChunkNoiseSamplerDensityInterpolator#wrapSample)
+                if (context instanceof NoisePosVanillaInterface vif && vif.getType() == EvalType.INTERPOLATION && NoiseChunk.this.interpolating) {
+                    int cellBlockX = context.blockX() - NoiseChunk.this.cellStartBlockX;
+                    int cellBlockY = context.blockY() - NoiseChunk.this.cellStartBlockY;
+                    int cellBlockZ = context.blockZ() - NoiseChunk.this.cellStartBlockZ;
+                    return NoiseChunk.this.fillingCell
+                        ? Mth.lerp3(
+                        (double) cellBlockX / NoiseChunk.this.cellWidth,
+                        (double) cellBlockY / NoiseChunk.this.cellHeight,
+                        (double) cellBlockZ / NoiseChunk.this.cellWidth,
+                        this.noise000,
+                        this.noise100,
+                        this.noise010,
+                        this.noise110,
+                        this.noise001,
+                        this.noise101,
+                        this.noise011,
+                        this.noise111
+                    )
+                        : this.value;
+                }
+                // DivineMC end - Density Function Compiler
                 return this.noiseFiller.compute(context);
             } else if (!NoiseChunk.this.interpolating) {
                 throw new IllegalStateException("Trying to sample interpolator outside the interpolation loop");
             } else {
                 return NoiseChunk.this.fillingCell
                     ? Mth.lerp3(
-                        (double)NoiseChunk.this.inCellX / NoiseChunk.this.cellWidth,
-                        (double)NoiseChunk.this.inCellY / NoiseChunk.this.cellHeight,
-                        (double)NoiseChunk.this.inCellZ / NoiseChunk.this.cellWidth,
+                        (double) NoiseChunk.this.inCellX / NoiseChunk.this.cellWidth,
+                        (double) NoiseChunk.this.inCellY / NoiseChunk.this.cellHeight,
+                        (double) NoiseChunk.this.inCellZ / NoiseChunk.this.cellWidth,
                         this.noise000,
                         this.noise100,
                         this.noise010,
diff --git a/net/minecraft/world/level/levelgen/NoiseRouter.java b/net/minecraft/world/level/levelgen/NoiseRouter.java
index 04890a6423432bf506a7465693000fcee4d66f98..a4eb9e9325003fadbeba12036f891cc000e67d10 100644
--- a/net/minecraft/world/level/levelgen/NoiseRouter.java
+++ b/net/minecraft/world/level/levelgen/NoiseRouter.java
@@ -4,23 +4,144 @@ import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.util.function.Function;
 
-public record NoiseRouter(
-    DensityFunction barrierNoise,
-    DensityFunction fluidLevelFloodednessNoise,
-    DensityFunction fluidLevelSpreadNoise,
-    DensityFunction lavaNoise,
-    DensityFunction temperature,
-    DensityFunction vegetation,
-    DensityFunction continents,
-    DensityFunction erosion,
-    DensityFunction depth,
-    DensityFunction ridges,
-    DensityFunction preliminarySurfaceLevel,
-    DensityFunction finalDensity,
-    DensityFunction veinToggle,
-    DensityFunction veinRidged,
-    DensityFunction veinGap
-) {
+// DivineMC start - Density Function Compiler - record converted to a class so it can carry the
+// precompiled finalDensity+Beardifier ("finalFinalDensity") and the uncompiled original router
+// (upstream MixinNoiseRouter + NoiseRouterExtension); accessors keep record naming
+public final class NoiseRouter implements com.ishland.c2me.opts.dfc.common.ducks.NoiseRouterExtension {
+    private final DensityFunction barrierNoise;
+    private final DensityFunction fluidLevelFloodednessNoise;
+    private final DensityFunction fluidLevelSpreadNoise;
+    private final DensityFunction lavaNoise;
+    private final DensityFunction temperature;
+    private final DensityFunction vegetation;
+    private final DensityFunction continents;
+    private final DensityFunction erosion;
+    private final DensityFunction depth;
+    private final DensityFunction ridges;
+    private final DensityFunction preliminarySurfaceLevel;
+    private final DensityFunction finalDensity;
+    private final DensityFunction veinToggle;
+    private final DensityFunction veinRidged;
+    private final DensityFunction veinGap;
+
+    private DensityFunction c2me$finalFinalDensity;
+    private NoiseRouter c2me$originalNoiseRouter;
+
+    public NoiseRouter(
+        DensityFunction barrierNoise,
+        DensityFunction fluidLevelFloodednessNoise,
+        DensityFunction fluidLevelSpreadNoise,
+        DensityFunction lavaNoise,
+        DensityFunction temperature,
+        DensityFunction vegetation,
+        DensityFunction continents,
+        DensityFunction erosion,
+        DensityFunction depth,
+        DensityFunction ridges,
+        DensityFunction preliminarySurfaceLevel,
+        DensityFunction finalDensity,
+        DensityFunction veinToggle,
+        DensityFunction veinRidged,
+        DensityFunction veinGap
+    ) {
+        this.barrierNoise = barrierNoise;
+        this.fluidLevelFloodednessNoise = fluidLevelFloodednessNoise;
+        this.fluidLevelSpreadNoise = fluidLevelSpreadNoise;
+        this.lavaNoise = lavaNoise;
+        this.temperature = temperature;
+        this.vegetation = vegetation;
+        this.continents = continents;
+        this.erosion = erosion;
+        this.depth = depth;
+        this.ridges = ridges;
+        this.preliminarySurfaceLevel = preliminarySurfaceLevel;
+        this.finalDensity = finalDensity;
+        this.veinToggle = veinToggle;
+        this.veinRidged = veinRidged;
+        this.veinGap = veinGap;
+    }
+
+    public DensityFunction barrierNoise() {
+        return this.barrierNoise;
+    }
+
+    public DensityFunction fluidLevelFloodednessNoise() {
+        return this.fluidLevelFloodednessNoise;
+    }
+
+    public DensityFunction fluidLevelSpreadNoise() {
+        return this.fluidLevelSpreadNoise;
+    }
+
+    public DensityFunction lavaNoise() {
+        return this.lavaNoise;
+    }
+
+    public DensityFunction temperature() {
+        return this.temperature;
+    }
+
+    public DensityFunction vegetation() {
+        return this.vegetation;
+    }
+
+    public DensityFunction continents() {
+        return this.continents;
+    }
+
+    public DensityFunction erosion() {
+        return this.erosion;
+    }
+
+    public DensityFunction depth() {
+        return this.depth;
+    }
+
+    public DensityFunction ridges() {
+        return this.ridges;
+    }
+
+    public DensityFunction preliminarySurfaceLevel() {
+        return this.preliminarySurfaceLevel;
+    }
+
+    public DensityFunction finalDensity() {
+        return this.finalDensity;
+    }
+
+    public DensityFunction veinToggle() {
+        return this.veinToggle;
+    }
+
+    public DensityFunction veinRidged() {
+        return this.veinRidged;
+    }
+
+    public DensityFunction veinGap() {
+        return this.veinGap;
+    }
+
+    @Override
+    public DensityFunction c2me$getFinalFinalDensity() {
+        return this.c2me$finalFinalDensity;
+    }
+
+    @Override
+    public void c2me$setFinalFinalDensity(DensityFunction densityFunction) {
+        this.c2me$finalFinalDensity = densityFunction;
+    }
+
+    @Override
+    public void c2me$setOriginalNoiseRouter(NoiseRouter originalNoiseRouter) {
+        this.c2me$originalNoiseRouter = originalNoiseRouter;
+    }
+
+    @Override
+    public NoiseRouter c2me$getOriginalNoiseRouter() {
+        return this.c2me$originalNoiseRouter;
+    }
+    // DivineMC end - Density Function Compiler
+
     public static final Codec<NoiseRouter> CODEC = RecordCodecBuilder.create(
         i -> i.group(
                 field("barrier", NoiseRouter::barrierNoise),
@@ -47,7 +168,7 @@ public record NoiseRouter(
     }
 
     public NoiseRouter mapAll(final DensityFunction.Visitor visitor) {
-        return new NoiseRouter(
+        NoiseRouter mapped = new NoiseRouter(
             this.barrierNoise.mapAll(visitor),
             this.fluidLevelFloodednessNoise.mapAll(visitor),
             this.fluidLevelSpreadNoise.mapAll(visitor),
@@ -64,5 +185,12 @@ public record NoiseRouter(
             this.veinRidged.mapAll(visitor),
             this.veinGap.mapAll(visitor)
         );
+        // DivineMC start - Density Function Compiler - propagate the compiled extras through visitors (upstream MixinNoiseRouter#postApply)
+        if (this.c2me$finalFinalDensity != null) {
+            mapped.c2me$finalFinalDensity = this.c2me$finalFinalDensity.mapAll(visitor);
+        }
+        mapped.c2me$originalNoiseRouter = this.c2me$originalNoiseRouter;
+        // DivineMC end - Density Function Compiler
+        return mapped;
     }
 }
diff --git a/net/minecraft/world/level/levelgen/RandomState.java b/net/minecraft/world/level/levelgen/RandomState.java
index e0c84dc7364e3a068f1325a45dee8880e4091fbe..750b1cbbe0bd5d252fa8a3ffeeb9fd7d607f89b6 100644
--- a/net/minecraft/world/level/levelgen/RandomState.java
+++ b/net/minecraft/world/level/levelgen/RandomState.java
@@ -107,6 +107,49 @@ public final class RandomState {
             this.router.ridges().mapAll(noiseFlattener),
             settings.spawnTarget()
         );
+
+        // DivineMC start - Density Function Compiler - compile the router and climate sampler to JVM bytecode (upstream MixinNoiseConfig)
+        if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.enableDensityFunctionCompiler) {
+            com.google.common.base.Stopwatch stopwatch = com.google.common.base.Stopwatch.createStarted();
+            it.unimi.dsi.fastutil.objects.Reference2ReferenceMap<DensityFunction, com.ishland.c2me.opts.dfc.common.ast.opto.OptoPasses.AstPair> optoCache = new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>();
+            it.unimi.dsi.fastutil.objects.Reference2ReferenceMap<DensityFunction, DensityFunction> tempCache = new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>();
+            DensityFunction finalFinalDensity = DensityFunctions.add(this.router.finalDensity(), DensityFunctions.BeardifierMarker.INSTANCE);
+            NoiseRouter original = this.router;
+            original.c2me$setFinalFinalDensity(finalFinalDensity);
+            this.router = new NoiseRouter(
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("barrier", this.router.barrierNoise(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("fluid_level_floodedness", this.router.fluidLevelFloodednessNoise(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("fluid_level_spread", this.router.fluidLevelSpreadNoise(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("lava", this.router.lavaNoise(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("temperature", this.router.temperature(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("vegetation", this.router.vegetation(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("continents", this.router.continents(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("erosion", this.router.erosion(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("depth", this.router.depth(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("ridges", this.router.ridges(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("preliminary_surface_level", this.router.preliminarySurfaceLevel(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("final_density", this.router.finalDensity(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("vein_toggle", this.router.veinToggle(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("vein_ridged", this.router.veinRidged(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("vein_gap", this.router.veinGap(), optoCache, tempCache)
+            );
+            this.router.c2me$setFinalFinalDensity(
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("final_final_density", finalFinalDensity, optoCache, tempCache)
+            );
+            this.router.c2me$setOriginalNoiseRouter(original);
+            this.sampler = new Climate.Sampler(
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("multiNoiseSampler_temperature", this.sampler.temperature(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("multiNoiseSampler_vegetation", this.sampler.humidity(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("multiNoiseSampler_continents", this.sampler.continentalness(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("multiNoiseSampler_erosion", this.sampler.erosion(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("multiNoiseSampler_depth", this.sampler.depth(), optoCache, tempCache),
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.compile("multiNoiseSampler_ridges", this.sampler.weirdness(), optoCache, tempCache),
+                this.sampler.spawnTarget()
+            );
+            stopwatch.stop();
+            org.slf4j.LoggerFactory.getLogger("DivineMC").info("Density function compilation finished in {}", stopwatch);
+        }
+        // DivineMC end - Density Function Compiler
     }
 
     public NormalNoise getOrCreateNoise(final ResourceKey<NormalNoise.NoiseParameters> noise) {
diff --git a/net/minecraft/world/level/levelgen/synth/BlendedNoise.java b/net/minecraft/world/level/levelgen/synth/BlendedNoise.java
index 90736b2dbd1bc6e67f1d00ca200b48c0a95e3d13..c6175ea210c97805cd4629f034d29e0fbe4bde33 100644
--- a/net/minecraft/world/level/levelgen/synth/BlendedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/BlendedNoise.java
@@ -179,4 +179,35 @@ public class BlendedNoise implements DensityFunction.SimpleFunction {
     public KeyDispatchDataCodec<? extends DensityFunction> codec() {
         return CODEC;
     }
+
+    // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream equality MixinInterpolatedNoiseSampler)
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        BlendedNoise that = (BlendedNoise) o;
+        return Double.compare(this.xzMultiplier, that.xzMultiplier) == 0
+            && Double.compare(this.yMultiplier, that.yMultiplier) == 0
+            && Double.compare(this.xzFactor, that.xzFactor) == 0
+            && Double.compare(this.yFactor, that.yFactor) == 0
+            && Double.compare(this.smearScaleMultiplier, that.smearScaleMultiplier) == 0
+            && java.util.Objects.equals(this.minLimitNoise, that.minLimitNoise)
+            && java.util.Objects.equals(this.maxLimitNoise, that.maxLimitNoise)
+            && java.util.Objects.equals(this.mainNoise, that.mainNoise);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+        result = 31 * result + java.util.Objects.hashCode(this.minLimitNoise);
+        result = 31 * result + java.util.Objects.hashCode(this.maxLimitNoise);
+        result = 31 * result + java.util.Objects.hashCode(this.mainNoise);
+        result = 31 * result + Double.hashCode(this.xzMultiplier);
+        result = 31 * result + Double.hashCode(this.yMultiplier);
+        result = 31 * result + Double.hashCode(this.xzFactor);
+        result = 31 * result + Double.hashCode(this.yFactor);
+        result = 31 * result + Double.hashCode(this.smearScaleMultiplier);
+        return result;
+    }
+    // DivineMC end - Density Function Compiler
 }
diff --git a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
index 6bc902e9f15b556a76c023c171ac489ba4939029..5db3cf351c91f71d14f858fd29af625e6e876c67 100644
--- a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -229,4 +229,27 @@ public final class ImprovedNoise {
     public void parityConfigString(final StringBuilder sb) {
         NoiseUtils.parityNoiseOctaveConfigString(sb, this.xo, this.yo, this.zo, this.p);
     }
+
+    // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream MixinPerlinNoiseSampler)
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ImprovedNoise that = (ImprovedNoise) o;
+        return Double.compare(this.xo, that.xo) == 0
+            && Double.compare(this.yo, that.yo) == 0
+            && Double.compare(this.zo, that.zo) == 0
+            && java.util.Arrays.equals(this.p, that.p);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+        result = 31 * result + java.util.Arrays.hashCode(this.p);
+        result = 31 * result + Double.hashCode(this.xo);
+        result = 31 * result + Double.hashCode(this.yo);
+        result = 31 * result + Double.hashCode(this.zo);
+        return result;
+    }
+    // DivineMC end - Density Function Compiler
 }
diff --git a/net/minecraft/world/level/levelgen/synth/NormalNoise.java b/net/minecraft/world/level/levelgen/synth/NormalNoise.java
index 868b96065b665c2653d5005bc91da8ef5f149211..5c312d89ff9f901001f0482d3d19b01ccd4ec5c7 100644
--- a/net/minecraft/world/level/levelgen/synth/NormalNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/NormalNoise.java
@@ -111,4 +111,25 @@ public class NormalNoise {
             this(firstOctave, Util.make(new DoubleArrayList(amplitudes), list -> list.add(0, firstAmplitude)));
         }
     }
+
+    // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream MixinDoublePerlinNoiseSampler)
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        NormalNoise that = (NormalNoise) o;
+        return Double.compare(this.valueFactor, that.valueFactor) == 0
+            && java.util.Objects.equals(this.first, that.first)
+            && java.util.Objects.equals(this.second, that.second);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+        result = 31 * result + Double.hashCode(this.valueFactor);
+        result = 31 * result + java.util.Objects.hashCode(this.first);
+        result = 31 * result + java.util.Objects.hashCode(this.second);
+        return result;
+    }
+    // DivineMC end - Density Function Compiler
 }
diff --git a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
index 36dbb7522462910f04c643ce4626490be1f0eccd..5c18646aa53b396c41a03185a43818d65cea00e9 100644
--- a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
@@ -242,4 +242,27 @@ public class PerlinNoise {
         sb.append("]");
         sb.append("}");
     }
+
+    // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream MixinOctavePerlinNoiseSampler)
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        PerlinNoise that = (PerlinNoise) o;
+        return Double.compare(this.lowestFreqValueFactor, that.lowestFreqValueFactor) == 0
+            && Double.compare(this.lowestFreqInputFactor, that.lowestFreqInputFactor) == 0
+            && java.util.Arrays.equals(this.noiseLevels, that.noiseLevels)
+            && java.util.Arrays.equals(this.amplitudes.toDoubleArray(), that.amplitudes.toDoubleArray());
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+        result = 31 * result + java.util.Arrays.hashCode(this.noiseLevels);
+        result = 31 * result + java.util.Arrays.hashCode(this.amplitudes.toDoubleArray());
+        result = 31 * result + Double.hashCode(this.lowestFreqValueFactor);
+        result = 31 * result + Double.hashCode(this.lowestFreqInputFactor);
+        return result;
+    }
+    // DivineMC end - Density Function Compiler
 }
diff --git a/net/minecraft/world/level/levelgen/synth/SimplexNoise.java b/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
index 5d9568ad86b573f89b42e8eec9f9143a961808a0..a65e61c446c24651539f96f0f4afb189c1b1e9fe 100644
--- a/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
@@ -191,4 +191,27 @@ public class SimplexNoise {
         double n3 = this.getCornerNoise3D(gi3, x3, y3, z3, 0.6);
         return 32.0 * (n0 + n1 + n2 + n3);
     }
+
+    // DivineMC start - Density Function Compiler - equality for AST deduplication (upstream MixinSimplexNoiseSampler)
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        SimplexNoise that = (SimplexNoise) o;
+        return Double.compare(this.xo, that.xo) == 0
+            && Double.compare(this.yo, that.yo) == 0
+            && Double.compare(this.zo, that.zo) == 0
+            && java.util.Arrays.equals(this.p, that.p);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+        result = 31 * result + java.util.Arrays.hashCode(this.p);
+        result = 31 * result + Double.hashCode(this.xo);
+        result = 31 * result + Double.hashCode(this.yo);
+        result = 31 * result + Double.hashCode(this.zo);
+        return result;
+    }
+    // DivineMC end - Density Function Compiler
 }
